pyomo/dae/tests/test_flatten.py

```diff
@@ -208,7 +293,997 @@ def c_rule(m, t):
+            elif len(sets) == 2 and sets[0] is m.time and sets[1] is m.time:
+                ref_data = {
+                    self._hashRef(Reference(m.v_tt)),
+                }
+                ref_data.update(self._hashRef(Reference(m.v_tst[:,x,:])) for x in S)
+                ref_data.update(self._hashRef(Reference(m.b.b2[:,x].v2[:,j])) for x, j in SC)
+                assert len(comps) == len(ref_data)
+                for comp in comps:
+                    self.assertIn(self._hashRef(comp), ref_data)
+            else:
+                raise RuntimeError()
+
+    def _model_2(self):
+        # A more simple model, but now with some higher-dimension sets
+        m = ConcreteModel()
+
+        normalize_index.flatten = False
+
+        m.d1 = Set(initialize=[1,2])
+        m.d2 = Set(initialize=[('a',1), ('b',2)])
+        m.dn = Set(initialize=[('c',3), ('d',4,5)], dimen=None)
+
+        m.v_2n = Var(m.d2, m.dn)
+        m.v_12 = Var(m.d1, m.d2)
+        m.v_212 = Var(m.d2, m.d1, m.d2)
+        m.v_12n = Var(m.d1, m.d2, m.dn)
+        m.v_1n2n = Var(m.d1, m.dn, m.d2, m.dn)
+
+        @m.Block(m.d1, m.d2, m.dn)
+        def b(b, i1, i2, i3):
+            b.v0 = Var()
+            b.v1 = Var(m.d1)
+            b.v2 = Var(m.d2)
+            b.vn = Var(m.dn)
+
+        normalize_index.flatten = True
+
+        return m
+
+    def test_flatten_m2_2d(self):
+        """
+        This test has some issues due to incompatibility between
+        slicing and `normalize_index.flatten==False`.
+        """
+        m = self._model_2()
+
+        sets = ComponentSet((m.d2,))
+        # need to set `flatten` to False here to properly access data,
+        # since model was created with `flatten == False`.
+        normalize_index.flatten = False
+        sets_list, comps_list = flatten_components_along_sets(m, sets, Var)
+
+        ref1 = Reference(m.v_2n[:,('c',3)])
+
+        ref_set = ref1._index._ref # _index is a _ReferenceSet
+
+        #next(ref_set._get_iter(ref_set._slice, ('a',1)))
+        # ^ This raises a somewhat cryptic error message.
+        # _ReferenceSet.__contains__ seems to be the culprit,
+        # which is called in validate_index for every __getitem__.
+        self.assertNotIn(('a',1), ref1._index._ref)
+        # ^ This does not seem to behave as expected...
+        # Reason is incompatibility with flatten==False.
+
+        self.assertEqual(len(sets_list), len(comps_list))
+        self.assertEqual(len(sets_list), 2)
+
+        #for sets, comps in zip(sets_list, comps_list):
+        #    if len(sets) == 1 and sets[0] is m.d2:
+        #        ref_data = {
+        #                self._hashRef(Reference(m.v_2n[:,('a',1)])),
+        #                self._hashRef(Reference(m.v_2n[:,('b',2)])),
+        #                }
+        #        # Cannot access the data of ^these references
+        #        for comp in comps:
+        #            self.assertIn(self._hashRef(comp), ref_data)
```

Comment:
Can this commented code be removed? At a higher level, is this test complete and doing what you expect? The comments seem to indicate that there may have been some issues debugging it.

Reply:
This test is basically not doing what is expected. The line
```python
self.assertNotIn(('a', 1), ref1._index._ref)
```
confirms this. `('a', 1)` should be a valid index. This is a problem with compatibility between slices and `normalize_index.flatten == False`. Basically a one-dimensional slice does not properly slice a higher-dimension set in this case. I'm not sure there is an easy fix. The commented code is what I would have expected to work.

Probably should just open an issue about this then remove this test?

Reply:
Thanks for the clarification. I think the best thing to do is to open an issue and then add a "TODO" comment in this test that references the issue. I think leaving the commented test for future reference is fine as long as we can follow the context of why it's there.

Reply:
Added comment and opened issue